### PR TITLE
Implement description required policy

### LINKED ISSUE #3469: Identify different Webauthn Tokens of a user
If WebAuthn gets more common, then a user might have several WebAuthn Tokens within privacyIDEA.

One on his Phone, on his tablet, on his notebook.
If the user looses one of these devices, it is not straightforward to know, which token was lost.

A solution would be be to force setting a description during enrollment by the user.

### PR DESCRIPTION
**Pull request by NNLemling**
closes #3469

### CODE REVIEW

File: privacyidea/lib/policy.py [hunk 1]
```diff
@@ -2117,6 +2118,17 @@ def get_static_policy_definitions(scope=None):
                 'type': 'int',
                 'desc': _('Limit the number of allowed tokens in a realm.'),
                 'group': GROUP.TOKEN},
+            ACTION.REQUIRE_DESCRIPTION: {
+                'type': 'str',
+                'desc': _('During the rollout process, this policy makes the '
+                          'description required for all selected token-types.'),
+                'group': GROUP.ENROLLMENT,
+                'multiple': True,
+                'value': ['4eyes', 'certificate', 'email', 'hotp', 'indexedsecret',
```
Comment (plettich) [inline privacyidea/lib/policy.py:2127]: You can just use [`get_token_types()`](https://github.com/privacyidea/privacyidea/blob/49f64d01b2bffeacabeac721a405d46d9a62725d/privacyidea/lib/config.py#L481) here

File: tests/test_api_lib_policy.py [hunk 2]
```diff
@@ -3342,6 +3343,30 @@ def test_60_increase_failcounter_on_challenge(self):
...
+        builder = EnvironBuilder(method='POST',
+                                 data={'user': "hans", "pass": "123456"},
+                                 headers={})
+        env = builder.get_environ()
+        req = Request(env)
+        req.User = User()
+
+        # Check that parameter is not set
+        req.all_data = {"type": "test_type"}
+        self.assertNotIn("require_description", req.all_data)
```
Comment (plettich) [inline tests/test_api_lib_policy.py:3356]: this test won't work if the `require_description` key is not added to `all_data` (see above).
We could create a test environment where the `description` parameter is missing just for the lib-tests.

File: privacyidea/api/lib/prepolicy.py [hunk 3]
```diff
@@ -2253,3 +2253,39 @@ def increase_failcounter_on_challenge(request=None, action=None):
...
+                             adminuser=adminuser,
+                             user=username,
+                             realm=realm,
+                             user_object=user_object).action_values(unique=False)
+
+    token_types = list(action_values.keys())
+    # Add parameter require_description
+    request.all_data["require_description"] = token_types
+    # check if description is required
+    if request.all_data.get("type") and request.all_data["type"] in token_types:
```
Comment (plettich) [inline privacyidea/api/lib/prepolicy.py:2289]: There is no check if the `description` is given in the parameters

File: tests/test_api_token.py [hunk 4]
```diff
@@ -2697,6 +2697,37 @@ def test_42_init_verify_sms_token(self):
...
+                                               "otpkey": self.otpkey,
+                                               "type": "hotp"},
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            result = res.json.get("result")
+            # a policyerror should be raised because hotp needs a description
+            self.assertEqual(result.get("error").get("message"),
+                             "Description required for hotp token.")
+
+        # try to enroll a totp token
```
Comment (plettich) [inline tests/test_api_token.py:2718]: It would be more useful to test if the hotp-token works when a description is given in the parameters.

**Review by plettich — CHANGES_REQUESTED** [review]

File: doc/policies/enrollment.rst [hunk 5]
```diff
@@ -840,3 +840,18 @@ is written to the orgnaizationalUnit (OU) of the request.
...
+
+
+.. _policy_description_required:
+
+description_required
+~~~~~~~~~~~~~~~~~~~~
+
+type: list
+
+To prevent tokens from becoming unidentifiable after a device loss. A description can
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/policies/enrollment.rst]: fixed in 0f971396dd9b42f2cce916cf6a902ed744eae788.

File: doc/policies/enrollment.rst [hunk 6]
```diff
@@ -840,3 +840,18 @@ is written to the orgnaizationalUnit (OU) of the request.
...
+be enforced with the "require_description policy". The desired token-types can be
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/policies/enrollment.rst:853]: fixed in 0f971396dd9b42f2cce916cf6a902ed744eae788.

File: doc/policies/enrollment.rst [hunk 7]
```diff
@@ -840,3 +840,18 @@ is written to the orgnaizationalUnit (OU) of the request.
...
+selected here. After setting up the policy, the selected token types can only be
+enrolled if a description is set.
+
+.. note:: In order to make the change effective in the webui as well, it is necessary to reload it.
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/policies/enrollment.rst]: fixed in 0f971396dd9b42f2cce916cf6a902ed744eae788.

File: privacyidea/api/lib/postpolicy.py [hunk 8]
```diff
@@ -589,13 +589,15 @@ def get_webui_settings(request, response):
                                                     user=loginname, realm=realm).action_values(unique=True)
         logout_redirect_url_pol = Match.generic(g, scope=SCOPE.WEBUI, action=ACTION.LOGOUT_REDIRECT,
                                                 user=loginname, realm=realm).action_values(unique=True)
+        require_description = Match.action_only(g, scope=SCOPE.ENROLL, action=ACTION.REQUIRE_DESCRIPTION).action_values(unique=False)
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/postpolicy.py]: fixed in 17f25e231ccc5f79535b1dc775b06e0cf17401a5.

File: privacyidea/api/lib/prepolicy.py [hunk 9]
```diff
@@ -2253,3 +2253,39 @@ def increase_failcounter_on_challenge(request=None, action=None):
...
+                             scope=SCOPE.ENROLL,
+                             adminrealm=adminrealm,
+                             adminuser=adminuser,
+                             user=username,
+                             realm=realm,
+                             user_object=user_object).action_values(unique=False)
+
+    token_types = list(action_values.keys())
+    # Add parameter require_description
+    request.all_data["require_description"] = token_types
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/prepolicy.py]: fixed in 342e843ee419d4fac37fa212c8fe9406b8c8d0c9.

File: privacyidea/api/lib/prepolicy.py [hunk 10]
```diff
@@ -2253,3 +2253,39 @@ def increase_failcounter_on_challenge(request=None, action=None):
...
+    # check if description is required
+    if request.all_data.get("type") and request.all_data["type"] in token_types:
+        log.warning("Missing description for {} token.".format(request.all_data["type"]))
+        raise PolicyError("Description required for {} token.".format(request.all_data["type"]))
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/prepolicy.py]: fixed in 342e843ee419d4fac37fa212c8fe9406b8c8d0c9.

File: privacyidea/lib/policy.py (hunk 1 above)
Comment (NNLemling) [reply privacyidea/lib/policy.py]: fixed in 79b20dd1280e6525c00936c96866e72094bde586.

File: privacyidea/lib/policy.py [hunk 11]
```diff
@@ -2117,6 +2118,14 @@ def get_static_policy_definitions(scope=None):
                 'type': 'int',
                 'desc': _('Limit the number of allowed tokens in a realm.'),
                 'group': GROUP.TOKEN},
+            ACTION.REQUIRE_DESCRIPTION: {
+                'type': 'str',
+                'desc': _('During the rollout process, this policy makes the '
+                          'description required for all selected token-types.'),
+                'group': GROUP.ENROLLMENT,
+                'multiple': True,
+                'value': list(get_token_types())},
```
Comment (plettich) [inline privacyidea/lib/policy.py:2127]: `get_token_types()` already returns a list.

**Review by plettich — CHANGES_REQUESTED** [review]

File: tests/test_api_lib_policy.py (hunk 2 above)
Comment (NNLemling) [reply tests/test_api_lib_policy.py]: fixed in 2ca66d18f3cc56625edbacac143bb58cb9a3ee07.

File: tests/test_api_token.py (hunk 4 above)
Comment (NNLemling) [reply tests/test_api_token.py:2729]: fixed in 2ca66d18f3cc56625edbacac143bb58cb9a3ee07.

File: privacyidea/lib/policy.py (hunk 1 above)
Comment (NNLemling) [reply privacyidea/lib/policy.py]: I accidentally wrapped get_token_types() in square brackets, which is why it didn't work for me without list().

File: privacyidea/lib/policy.py (hunk 11 above)
Comment (NNLemling) [reply privacyidea/lib/policy.py]: You are right!

I accidentally wrapped get_token_types() in square brackets, which is why it didn't work for me without list().

fixed in f7ace1114f81c902ec128e78557a3554fb2001c6.

File: privacyidea/lib/policy.py [hunk 12]
```diff
@@ -2117,6 +2118,14 @@ def get_static_policy_definitions(scope=None):
                 'type': 'int',
                 'desc': _('Limit the number of allowed tokens in a realm.'),
                 'group': GROUP.TOKEN},
+            ACTION.REQUIRE_DESCRIPTION: {
+                'type': 'str',
+                'desc': _('During the rollout process, this policy makes the '
+                          'description required for all selected token-types.'),
```
Comment (cornelinux) [inline privacyidea/lib/policy.py:2124]: I think we usually write tokentype in one word:

"token-types" -> "tokentypes".

**Review by cornelinux — CHANGES_REQUESTED** [review]
Mostly docs issues.

however, please check my thoughts about the default tokentype.

File: privacyidea/api/lib/prepolicy.py [hunk 13]
```diff
@@ -2253,3 +2253,35 @@ def increase_failcounter_on_challenge(request=None, action=None):
...
+    action_values = Match.generic(g, action=ACTION.REQUIRE_DESCRIPTION,
+                             scope=SCOPE.ENROLL,
+                             adminrealm=adminrealm,
+                             adminuser=adminuser,
+                             user=username,
+                             realm=realm,
+                             user_object=user_object).action_values(unique=False)
+
+    token_types = list(action_values.keys())
+    if request.all_data.get("type") and request.all_data["type"] in token_types:
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/prepolicy.py:2284]: A specific tokentype is always provide.

File: doc/policies/enrollment.rst [hunk 14]
```diff
@@ -840,3 +840,17 @@ is written to the orgnaizationalUnit (OU) of the request.
...
+
+require description
+~~~~~~~~~~~~~~~~~~~~
+
+type: list
+
+To prevent tokens from becoming unidentifiable after a device loss, a description can
+be enforced with the "require_description policy". The desired token-types can be
+selected here. After setting up the policy, the selected token types can only be
+enrolled if a description is set.
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/policies/enrollment.rst]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

File: privacyidea/api/lib/prepolicy.py [hunk 15]
```diff
@@ -2253,3 +2253,35 @@ def increase_failcounter_on_challenge(request=None, action=None):
...
+
+
+def require_description(request=None, action=None):
+    """
+    Pre Policy
+    This checks if a description is required to roll out a specific token.
+    scope=SCOPE.ENROLL, action=REQUIRE_DESCRIPTION
+
+    An exception is raised, if the token-types specified in the
+    REQUIRE_DESCRIPTION policy match the token to be rolled out.
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/prepolicy.py]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

File: tests/test_api_token.py [hunk 16]
```diff
@@ -2697,6 +2697,48 @@ def test_42_init_verify_sms_token(self):
...
+        # set require_description policy with value = 'hotp'
+        set_policy(name="require_description",
+                   scope=SCOPE.ENROLL,
+                   action=["{0!s}=hotp".format(ACTION.REQUIRE_DESCRIPTION)])
+        # try to enroll a hotp token
+        with self.app.test_request_context('/token/init',
+                                           method='POST',
+                                           data={
+                                               "otpkey": self.otpkey,
+                                               "type": "hotp"},
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply tests/test_api_token.py:2709]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

File: privacyidea/api/lib/prepolicy.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/api/lib/prepolicy.py]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

File: privacyidea/lib/policy.py (hunk 12 above)
Comment (NNLemling) [reply privacyidea/lib/policy.py]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

File: tests/test_api_lib_policy.py [hunk 17]
```diff
@@ -3342,6 +3343,30 @@ def test_60_increase_failcounter_on_challenge(self):
...
+        env = builder.get_environ()
+
+        # Set policy
+        set_policy(name="require_description",
+                   scope=SCOPE.ENROLL,
+                   action=["{0!s}=hotp".format(ACTION.REQUIRE_DESCRIPTION)])
+
+        req = Request(env)
+        # No description is set
+        req.all_data = {"type": "hotp"}
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply tests/test_api_lib_policy.py]: fixed in 38630095c78fff51b7fda94e221140375e04ac48.

**Review by cornelinux — APPROVED** [review]

**Pull request merged**